### DISCUSSION TITLE
recent_error utility function

### DISCUSSION THREAD

**Discussion by 1015onasaturdaynight** [nanodbc/nanodbc#310] [General]
Responding to the reference to the recent_error utility here:

https://github.com/nanodbc/nanodbc/commit/3bddef90738b0bb23fa6032c40272850b8b9a952

> It seems the "Violation of PRIMARY KEY constraint" error is
accompanied/concluded with "The statement has been terminated"
as there is a hierarchy or list of diagnostics collected by
the `recent_error` utility function:
https://github.com/nanodbc/nanodbc/blob/946084a9a2a7a34429390c23c56bee2be4a2da6d/nanodbc/nanodbc.cpp#L391-L441

I suspect this is related to an incorrect error return I encountered while I was adding Oracle tests.

When I call the unsupported function SQLTables SQL_ALL_CATALOGS via ODBC directly, the call returns (expected):

```
ret = SQLTables(
    stmt,
    (SQLCHAR*)SQL_ALL_CATALOGS,
    1,
    (SQLCHAR*)"",
    0,
    (SQLCHAR*)"",
    0,
    (SQLCHAR*)"",
    0);
```

```
native error code: 0
sqlstate: HYC00
detailed msg: [Oracle][ODBC]Optional feature not implemented.
```

However, when calling via nanodbc, the call returns:

```
auto names = catalog.list_catalogs();
```

```
native error code: 0
sqlstate: HYT00
detailed msg: (redacted)/src/3rdparty/nanodbc/2.14.0/nanodbc/nanodbc.cpp:4981: HYT00: [Oracle][ODBC]Timeout expired.                               
```

**Comment by mloskot** [comment]
When you are calling the `list_catalogs`, you should be able to step through the `recent_error` and observe if there is one or more diagnostics available (see also https://github.com/nanodbc/nanodbc/issues/315). It is possible, the `Optional feature not implemented` is 'swallowed' and omitted in the final error data collected.
Sorry, I currently don't have access to Oracle to test it myself.

**Comment by 1015onasaturdaynight** [comment]
After much effort to confirm there's only a single error returned by the Oracle driver... this ends up being the result of building/running nanodbc with SQL_OV_ODBC3_80. Noting this here for Oracle's next victim.

With SQL_OV_ODBC3_80, Oracle returns: ```HYT00: [Oracle][ODBC]Timeout expired.```

With SQL_OV_ODBC3, Oracle returns: ```HYC00: [Oracle][ODBC]Optional feature not implemented.```

Of course, handling the possibility of nested exceptions (https://github.com/nanodbc/nanodbc/issues/315) may still be warranted.

**Reply by mloskot** [reply]
I've never used nanodbc (or ODBC in general) with any Oracle drivers, so I had no chance to encounter that behaviour.
